### PR TITLE
[RFC] gossiper: apply updates in batches

### PR DESCRIPTION
**Pull request by denesb**
Currently the gossiper applies each message as it lands. The apply
process splits the message into per-endpoint states and applies these
states individually. This has a limited concurrency of 100, further
updates are queued on the semaphore. In practice this concurrency can be
reduced further as each update locks the endpoint it is updating, so if
there are multiple concurrent updates for the same endpoint, only one is
processed, the rest waits on the lock.
Updates are replicated to other shards as they are applied. This has
negligible impact usually. But as the number of nodes (and therefore the
number of messages to process) and number of shards scales up, the
chance of some shard lagging behind with applying the updates increases.
Applying the updates on remote shards is very little amount of work, but
the entire process is very sensitive to latency: the latency of applying
the message is that of the slowest shard replicating the update. It is
particularly bad if one of the shards is loaded with a process that is
violating its task-quota for an extended period of time. This means that
even though the gossip scheduling group still gets it runtime, it will
get it as little as just a few times per second, meaning that only a few
endpoint updates can be replicated per second. In a large cluster where
a single message can contain dozens of endpoint state updates, this can
lead to unprocessed messages queueing up unbounded, leading to OOM.

This patch attempts to alleviate this problem by squashing all
unprocessed updates into a single staging state. Applying the state is
moved into a separate loop, which applies the entire state in
iterations. While the loop is in an iteration, any incoming updates are
applied to the staging state, after which the loop is notified that
there is new state to process.

This patch is a very rough sketch of the idea. Many things are not
addressed:
* Intricacies of applying states;
* Error handling;
* Error propagation;
* Waiting for individual messages;

The goal of this patch is to start a discussion on this proposal based
on real code, without expending large amount of effort on correctness
and polishing on an idea that is potentially fundamentally flawed.

Fixes: https://github.com/scylladb/scylla/issues/10967

### CODE REVIEW

**Comment by denesb** [comment]
Not using the DRAFT PR feature of GH as it insists on not notifying reviewers.

File: gms/gossiper.cc [hunk 1]
```diff
@@ -654,6 +666,17 @@ future<> gossiper::apply_state_locally(std::map<inet_address, endpoint_state> ma
...
     logger.debug("apply_state_locally() took {} ms", std::chrono::duration_cast<std::chrono::milliseconds>(
         std::chrono::steady_clock::now() - start).count());
+  }
+}
+
+future<> gossiper::apply_state_locally(std::map<inet_address, endpoint_state> map) {
+    for (auto& entry : map) {
+        //FIXME: this probably has to be much more sophisticated than this, checking generations, etc.
+        //Q: are state updates always full overwrites? Do we ever have to do some merging?
+        _staging_endpoint_state_map[entry.first] = std::move(entry.second);
```
Comment (asias) [inline gms/gossiper.cc:676]: The endpoint_state to be applied is only partial. So the code here will overwrite any previous collected states.

**Comment by asias** [comment]
One issue with the merge thing I can think of: 

Before all versions of the application states are applied so all the on_change callback for the states are called. 
After only the latest application state is applied and only the latest on_change callback is called.

This is problematic. For instance, the node gets a STATUS = BOOTSTRAP (v1), then later it gets STATUS = NORMAL (v2). The STATUS = BOOTSTRAP is queued. Before the node has a chance to apply STATUS = BOOTSTRAP , it gets the STATUS = NORMAL.  In this case, the important callback for BOOTSTRAP will be missed.

**Comment by asias** [comment]
I think we do not really need all the complexity here. 

With a33c370f9a3, we now  have should_count_as_msg_processing.

We can do something like the following

if (!should_count_as_msg_processing(ep_state_map)  && _apply_state_locally_semaphore.waiters() > 100) {
      drop ep_state_map and skip apply the state_map
}

This way, those frequent but non critical state can be dropped.

File: gms/gossiper.cc (hunk 1 above)
Comment (asias) [reply gms/gossiper.cc:676]: The merge here is actually going to be pretty complicated. We need to consider both version and generations. This is going to replicate lots of the actual apply logic.

**Comment by denesb** [comment]
> One issue with the merge thing I can think of:
> 
> Before all versions of the application states are applied so all the on_change callback for the states are called. After only the latest application state is applied and only the latest on_change callback is called.
> 
> This is problematic. For instance, the node gets a STATUS = BOOTSTRAP (v1), then later it gets STATUS = NORMAL (v2). The STATUS = BOOTSTRAP is queued. Before the node has a chance to apply STATUS = BOOTSTRAP , it gets the STATUS = NORMAL. In this case, the important callback for BOOTSTRAP will be missed.

I think this is a deal-breaker for this idea indeed. We could deal with the complex merging, but it turns out we don't need to, it is simpler to just drop low-value messages.

**Comment by denesb** [comment]
Pulling the plug on this one, in favor of https://github.com/scylladb/scylla/pull/11039.

**Pull request closed without merging**